src/lib/services/client-metrics/last-seen/last-seen-store.ts

```diff
@@ -0,0 +1,75 @@
+import EventEmitter from 'events';
+import { LogProvider, Logger } from '../../../logger';
+import { DB_TIME } from '../../../metric-events';
+import { Db } from '../../../server-impl';
+import metricsHelper from '../../..//util/metrics-helper';
+import { LastSeenInput } from './last-seen-service';
+import { ILastSeenStore } from './types/last-seen-store-type';
+
+// const LAST_SEEN_AT_COLUMNS = ['feature_name', 'environment', 'last_seen_at'];
+const TABLE = 'last_seen_at_metrics';
+
+export interface FeaturesTable {
+    feature_name: string;
+    last_seen_at: Date;
+    environment: string;
+}
+
+export default class LastSeenStore implements ILastSeenStore {
+    private db: Db;
+
+    private logger: Logger;
+
+    private timer: Function;
+
+    constructor(db: Db, eventBus: EventEmitter, getLogger: LogProvider) {
+        this.db = db;
+        this.logger = getLogger('last-seen-store.ts');
+        this.timer = (action) =>
+            metricsHelper.wrapTimer(eventBus, DB_TIME, {
+                store: 'last-seen-environment-store',
+                action,
+            });
+    }
+
+    async exists(name: string): Promise<boolean> {
+        const result = await this.db.raw(
+            'SELECT EXISTS (SELECT 1 FROM features WHERE name = ?) AS present',
+            [name],
+        );
+        const { present } = result.rows[0];
+        return present;
+    }
+
+    async setLastSeen(data: LastSeenInput[]): Promise<void> {
+        const now = new Date();
+
+        try {
+            const inserts = data.map((item) => {
+                return {
+                    feature_name: item.featureName,
+                    environment: item.environment,
+                    last_seen_at: now,
+                };
+            });
+
+            const batchSize = 1000;
+
+            await this.db.transaction(async (trx) => {
```

Comment:
I'd do the transaction around the service (in the controller) as we do for all new code.

Reply:
Since it's a non critical data I'm wondering if having transactions even makes sense here. I'd consider sacrificing correctness for performance.

Reply:
Agree, done